### PR TITLE
Fix a dangling pointer expose to PDI

### PR DESCRIPTION
**Pull request by jbigot**
Also use PDI in heat_equation to test it

### CODE REVIEW

**Comment by tpadioleau** [comment]
Is this ready ?

**Comment by jbigot** [comment]
It's better than what we had before, but does not cleanly solve the issue. I'd merge as there is no downside, but keep the issue open

**Comment by tpadioleau** [comment]
I can merge but what would be the remaining issue ? The issue was about dangling pointers and this seems to fix it ?
We may have one remaining issue about the second `with` function for Arithmetic types. If a rvalue is passed, i think we end up in the same situation.

File: include/ddc/pdi.hpp [hunk 1]
```diff
@@ -41,11 +47,17 @@ class PdiEvent
         static_assert(
                 !(access & PDI_IN) || (chunk_default_access_v<BorrowedChunk> & PDI_IN),
                 "Invalid access for constant data");
+        auto allocator = std::pmr::polymorphic_allocator<size_t>(&m_metadata);
         auto extents = detail::array(data.domain().extents());
-        size_t rank = extents.size();
+        size_t& rank = *allocator.allocate(1);
+        rank = data.domain().extents().size();
```
Comment (tpadioleau) [inline include/ddc/pdi.hpp:53]:
```suggestion
        rank = extents.size();
```

File: include/ddc/pdi.hpp [hunk 2]
```diff
@@ -41,11 +47,17 @@ class PdiEvent
...
         PDI_share((name + "_rank").c_str(), &rank, PDI_OUT);
         m_names.push_back(name + "_rank");
-        PDI_share((name + "_extents").c_str(), extents.data(), PDI_OUT);
+        PDI_share(
+                (name + "_extents").c_str(),
+                std::vector<size_t, decltype(allocator)>(extents.begin(), extents.end(), allocator)
```
Comment (tpadioleau) [inline include/ddc/pdi.hpp:58]:
```suggestion
                std::pmr::vector<size_t>(extents.begin(), extents.end(), allocator)
```

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: MIT
 
 add_executable(heat_equation heat_equation.cpp)
-target_link_libraries(heat_equation PUBLIC DDC::DDC)
+target_link_libraries(heat_equation PUBLIC DDC::DDC DDC::PDI_Wrapper)
```
Comment (tpadioleau) [inline examples/CMakeLists.txt:4]: Shall we disable this example if `PDI_Wrapper` is not enabled ?

File: examples/heat_equation.cpp [hunk 4]
```diff
@@ -65,13 +65,19 @@ void display(double time, ChunkType temp)
...
+            .and_with("mean_temp", mean_temp)
+            .and_with("temp_slice", temp_slice);
 }
 //! [display]
 
 
 //! [main-start]
 int main(int argc, char** argv)
 {
+    PDI_init(PC_parse_string(""));
```
Comment (tpadioleau) [inline examples/heat_equation.cpp:80]: Does it leak when the string is empty ?

Comment (jbigot) [reply examples/heat_equation.cpp:80]: Yes, but is a leak that happens only once really a leak? :)

File: examples/CMakeLists.txt (hunk 3 above)
Comment (jbigot) [reply examples/CMakeLists.txt:4]: True... Maybe `#ifdef` the PDI part

File: examples/heat_equation.cpp (hunk 4 above)
Comment (tpadioleau) [reply examples/heat_equation.cpp:80]: I agree maybe not a true leak but still, it triggers the sanitizers. For this reason I would be in favor of calling the destructor of the object. It also shows the correct usage of the paraconf library.

File: examples/CMakeLists.txt (hunk 3 above)
Comment (tpadioleau) [reply examples/CMakeLists.txt:4]: Can you also add a if in the CMakeLists.txt

**Review by tpadioleau — APPROVED** [review]
At some point we should also add a guard about the `Arithmetic` case. It is not critical because a `PdiEvent` object is usually being destroyed after all the temporaries that are passed to the `with`/`and_with` functions.

**Pull request merged**